pkg/cloudevents/transport/x/example_using_test.go

```diff
@@ -0,0 +1,128 @@
+package x_test
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"strconv"
+
+	"github.com/cloudevents/sdk-go/pkg/cloudevents"
+	"github.com/cloudevents/sdk-go/pkg/cloudevents/client"
+)
+
+var count = 3 // Example ends after this many events.
+
+// The sender uses the cloudevents.Client API, not the transport APIs directly.
+func runSender(w io.Writer) error {
+	c, err := client.New(NewExTransport(nil, w))
+	if err != nil {
+		return err
+	}
+	for i := 0; i < count; i++ {
+		e := cloudevents.New()
+		e.SetType("example.com/event")
+		e.SetSource("example.com/source")
+		e.SetID(strconv.Itoa(i))
+		if err := e.SetData(fmt.Sprintf("hello %d", i)); err != nil {
+			return err
+		}
+		if _, err := c.Send(context.TODO(), e); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// The receiver uses the cloudevents.Client API, not the transport APIs directly.
+func runReceiver(r io.Reader) error {
+	i := 0
+	process := func(e cloudevents.Event) error {
+		fmt.Printf("%s\n", e)
+		i++
+		if i == count {
+			return io.EOF
+		}
+		return nil
+	}
+	if c, err := client.New(NewExTransport(r, nil)); err != nil {
+		return err
+	} else {
+		return c.StartReceiver(context.TODO(), process)
+	}
+}
+
+// The intermediary receives events and forwards them to another
+// process using ExReceiver and ExSender directly.
+//
+// By forwarding a transport.Message instead of a cloudevents.Event,
+// it allows the transports to avoid un-necessary decoding of
+// structured events, and to exchange delivery status between reliable
+// transports. Even transports using different protocols can ensure
+// reliable delivery.
+//
+func runIntermediary(r io.Reader, w io.WriteCloser) error {
+	for {
+		receiver := NewExReceiver(r)
+		sender := NewExSender(w)
+		for i := 0; i < count; i++ {
+			if m, err := receiver.Receive(context.TODO()); err != nil {
```

Comment:
Oh, I see, the http transport could send and the resp could be queued and ready for the next Receive call.

Reply:
Yep - key thing here is how to present the request/response correlation in the API. I'd like it not to be the first thing the users see, or perceived as the "normal" way of sending events, but I think we can find a way to provide it as an optional feature for transports (like HTTP) that are RCP underneath.

> Moving Message out of the transport also leads to interesting developments around sharing codec that are very common in strategy, like binary mode in AMQP, PubSub and HTTP all use the equivalent of headers to send the data, the Key value is different in each, but the strategy is the same...

Agreed, e.g. the `if Structured() else Event()` dance could be part of a generic library usable by all transports that allow structured tunneling.  Also the correct handling of different QoS options definitely should be captured generically, each transport just calls some relevant functions when it gets it's transport-specific ack indicator.